ikarus/solver/nonlinearsolver/nonlinearsolverfactory.hh

```diff
@@ -8,14 +8,54 @@
 
 #pragma once
 
+#include <type_traits>
 #include <utility>
 
+#include <dune/common/float_cmp.hh>
+
 #include <ikarus/assembler/dirichletbcenforcement.hh>
 #include <ikarus/utils/defaultfunctions.hh>
 #include <ikarus/utils/differentiablefunctionfactory.hh>
+#include <ikarus/utils/functionhelper.hh>
 
 namespace Ikarus {
 
+namespace Impl {
+  struct IDBCForceFunction
+  {
+    template <typename A>
+    auto operator()(const A& assembler) {
+      return [&]<typename D>(const D&) { return utils::obtainForcesDueToIDBC(assembler); };
+    }
+  };
+
+  template <typename CT, typename A, typename S>
+  auto updateFunctor(const A& assembler, const S& setting) {
+    return [&]<typename D, typename C>(D& x, const C& b) {
+      if constexpr (not std::is_same_v<C, utils::SyncFERequirements>) {
+        // the right-hand side is reduced
+        if (assembler->dBCOption() == DBCOption::Reduced and assembler->reducedSize() == b.size()) {
+          setting.updateFunction(x, assembler->createFullVector(b));
+        } else if (assembler->reducedSize() == x.size() and
+                   assembler->size() == b.size()) // the right-hand side is full but x is reduced
+        {
+          setting.updateFunction(x, assembler->createReducedVector(b));
+        } else
+          setting.updateFunction(x, b);
+      } else { // updates due to inhomogeneous bcs
+        if constexpr (requires { x.parameter(); }) {
```

Comment:
Hey @henrij22, I tried to simplify things here further using the commit https://github.com/ikarus-project/ikarus/commit/f0cb2a71c3c57afe0f9a3e743332ca0826fc1d82#diff-5ad4369a8e8d996635743b1b107846b8ac9f261e9e43ee6645bb81e04921da4c by implementing `syncFERequirement` function, but probably there is still some copy happeneing leading in `NaN` for arc length control only. So I reverted this back but simplified `IDBCForceFunction`. I don't know how to fix it further

Reply:
Also now assembler gives only a const ref and not a mutable requirement and hence it is difficult to use `syncFERequirement`

Reply:
Yes, it is there in my commit, but it broke something else and so for the moment I only reverted the changes related to the assembler and `syncFERequirement`